File: leetcode/tree/lowest_common_ancestor_bst.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	parent_val, p_val, q_val := root.Val, p.Val, q.Val
	if p_val > parent_val && q_val > parent_val {
		return lowestCommonAncestor(root.Right, p, q)
	} else if p_val < parent_val && q_val < parent_val {
		return lowestCommonAncestor(root.Left, p, q)
	} else {
		return root
	}
}

func main() {
	root := &TreeNode{6, nil, nil}
	root.Left = &TreeNode{2, nil, nil}
	root.Right = &TreeNode{8, nil, nil}
	root.Right.Left = &TreeNode{7, nil, nil}
	root.Right.Right = &TreeNode{9, nil, nil}
	root.Left.Left = &TreeNode{0, nil, nil}
	root.Left.Right = &TreeNode{4, nil, nil}
	root.Left.Right.Left = &TreeNode{3, nil, nil}
	root.Left.Right.Right = &TreeNode{5, nil, nil}
	q := &TreeNode{2, nil, nil}
	p := &TreeNode{4, nil, nil}
	res := lowestCommonAncestor(root, p, q)
	fmt.Println(res.Val)
}

File: leetcode/tree/lowest_common_ancestor_bst.py
from common import TreeNode
from typing import List

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        parent_val, p_val, q_val = root.val, p.val, q.val
        if p_val > parent_val and q_val > parent_val:
            return self.lowestCommonAncestor(root.right, p, q)
        elif p_val < parent_val and q_val < parent_val:
            return self.lowestCommonAncestor(root.left, p, q)
        else:
            return root

if __name__ == "__main__":
    root = TreeNode(6)
    root.left = TreeNode(2)
    root.right = TreeNode(8)
    root.left.left = TreeNode(2)
    root.right.right = TreeNode(8)
    root.left.left.left = TreeNode(0)
    root.left.left.right = TreeNode(4)
    root.left.left.right.left = TreeNode(3)
    root.left.left.right.right = TreeNode(5)
    root.right.left = TreeNode(7)
    root.right.right = TreeNode(9)
    p = TreeNode(2)
    q = TreeNode(4)
    res = Solution().lowestCommonAncestor(root, p, q)
    print(res.val)